sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsEnvironment.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.metrics;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Manages and provides the metrics container associated with each thread.
+ *
+ * <p>The runner should create {@link MetricsContainer} for each context in which metrics are
+ * reported (by step and name) and call {@link #setMetricsContainer} before invoking any code that
+ * may update metrics within that step.
+ * <p>The runner should call {@link #unsetMetricsContainer} (or {@link #setMetricsContainer} back to
+ * the previous value) when exiting code that set the metrics container.
+ */
+public class MetricsEnvironment {
```

Comment:
You should include a comment that users should never interact with this directly. Potentially this could be in a similar kind of state as `util`, where we don't generate any documentation for the runner-based utilities that have to exist in the SDK for visibility reasons.